File: match.php
<?php

$CPEs = file($argv[1], FILE_IGNORE_NEW_LINES);
$URLs = file($argv[2], FILE_IGNORE_NEW_LINES);
$license = "NOLICENSE";
if (isset($argv[3])) {
    $license = $argv[3];
}

$cpeUrlArray = array();
$cpeArray = array();

foreach ($CPEs as $cpe) {
    $check1 = explode(":", $cpe);

    $componentName = $check1[4];
    if ($check1[6] == "*" || $check1[6] == "-") {
        $ver = $check1[5];
        $ver = str_replace("alpha", "alpha.", $ver);
        $ver = str_replace("beta", "beta.", $ver);
        $ver = str_replace("milestone", "m", $ver);
        $ver = str_replace("-", ".", $ver);
        $ver = str_replace("-", ".", $ver);
        $ver = str_replace(".dev.", "", $ver);
        $ver = str_replace("rc", "rc.", $ver);
        /*$ver = str_replace("rc", "rc.", $ver);
        $ver = str_replace("1.0.0", "1.0", $ver);
        $ver = str_replace("1.1.0", "1.1", $ver);
        $ver = str_replace("1.2.0", "1.2", $ver);
        $ver = str_replace("2.0.0", "2.0", $ver);
        $ver = str_replace("2.1.0", "2.1", $ver);
        $ver = str_replace("2.2.0", "2.2", $ver);*/
        //$ver = str_replace($replaceThis, $withThis, $ver);
    } else {
        $ver = "$check1[5].$check1[6]";
        $ver = str_replace("alpha", "alpha.", $ver);
        $ver = str_replace("beta", "beta.", $ver);
        $ver = str_replace("milestone", "m", $ver);
        $ver = str_replace("-", ".", $ver);
        $ver = str_replace(":", ".", $ver);
        $ver = str_replace("rc", "rc.", $ver);
        $ver = str_replace(".dev.", "", $ver);
        /*$ver = str_replace("1.0.0", "1.0", $ver);
        $ver = str_replace("1.1.0", "1.1", $ver);
        $ver = str_replace("1.2.0", "1.2", $ver);
        $ver = str_replace("2.0.0", "2.0", $ver);
        $ver = str_replace("2.1.0", "2.1", $ver);
        $ver = str_replace("2.2.0", "2.2", $ver);*/
        //$ver = str_replace($replaceThis, $withThis, $ver);
        //$ver = str_replace(".rc", "rc", $ver);
    }
    foreach ($URLs as $url) {
        $origURL = $url;
        $url = strtolower($url);
        $array1 = explode("/", $url);
        $replaceThis = ["$componentName-", "miniupnpd_", "release-", "-release", "_", "v", "alpha", "beta", "rc", "ddpre", "beta1111", "beta2222", "beta311", "beta411", ".zip", ".exe", ".tar.gz", ".dmg", ".bin", ".win-amd64-py2.7", "-", ".win32-py2.7", ".win.x64.portable", ".win.x64.", ".tar.bz2", ".win32", ".win64", ".win.x86.portable", ".win.x86", ".final", "linux-", "e2fsprogs.", "hdf5.", "windows", "elasticsearch-oss.", "elasticsearch.", ".release", ".final", "platform.", ".release", "xwikiplatform", "elefant.", ".stable", "rrrrc.", "betaaaa", "rrc", ".msi", ".tar.xz", ".x64", ".x32",".tgz"];
        $withThis = ["", "", "", "", ".", "", "alpha.", "beta.", "rc.", ".pre.", ".beta.1", ".beta.2", "beta.3", "beta.4", "", "", "", "", "", "", ".", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "beta.", "rc.", "", "", "", "",""];
        if (isset($array1[7])) {
            //if ($array1[5] != ""){
            $urlVer = str_replace($replaceThis, $withThis, $array1[7]);
            //  }

        } else {
            $urlVer = str_replace($replaceThis, $withThis, end($array1));
        }

        $x = explode(".", $ver);
        $y = explode(".", $urlVer);
        $aa = strlen($urlVer);
        $ab = strlen($ver);

        //$ver = $ver.".0";
        
        // if (count($x) == 2) {
        //     if ($ver !== $urlVer) {
        //         $ver = $ver . ".0";
        //     }
        // }
        // print "$ver|$urlVer\n";
        if ($aa == $ab) {
            if ($ver == $urlVer && strlen(substr(strrchr($ver, "."), 1)) == strlen(substr(strrchr($urlVer, "."), 1))) {
                //if (strpos($urlVer,$ver) !== False ) {
                //print "$ver|$urlVer\n";

                $cpeUrl = $cpe . $origURL;
                array_push($cpeUrlArray, $cpeUrl);
                if (!in_array($cpe, $cpeArray)) {
                    array_push($cpeArray, $cpe);
                    //break;
                }
            }
        }
    }
}

foreach ($cpeUrlArray as $cpeResult) {
    if (substr($cpeResult, -1) !== "/") {
        print($cpeResult . "|" . $license . "\n");
    }
}
foreach ($CPEs as $cpe) {
    if (!in_array($cpe, $cpeArray)) {
        print "$cpe" . "NOURL|" . $license . "\n";
    }
}

File: match.py
import sys

def open_file(file):
    """
    Open file function
    """
    with open(file, encoding="ISO-8859-1") as opened_file:
        lines = opened_file.read().splitlines()
        return lines

REPLACE_CPE = ["alpha", "beta", "rc", "milestone", "pre"]
REPLACE_GIT = ["alpha", "beta", "rc", "milestone", "pre"]
UNUSUAL_GIT = ["a","b"]
CPE_URL = []
CPE_NOURL = []
CPE_ARRAY = []
try:
    CPE_LIST = open_file(sys.argv[1])
    GIT_LIST = open_file(sys.argv[2])
except IndexError:
    print(
        "Something is not right! Use the syntax :\npython3 match.py CPE-LIST GIT-LIST LICENSE-NAME STRING-THAT-YOU-NEED-TO-MATCH"
    )
    exit()
except FileNotFoundError:
    print("File not found!")
    exit()
except:
    print("Something went wrong! Mistakes were made!")
    exit()
try:
    LICENSE = sys.argv[3]
except:
    LICENSE = "NOLICENSE"

try:
    manual_component = sys.argv[4]
except:
    manual_component = ""        

for cpe in CPE_LIST:
    if cpe != "":
        organization = cpe.split(":")[3]
        component = cpe.split(":")[4]
        version = cpe.split(":")[5]
        subversion = cpe.split(":")[6]
        if subversion == "*" or subversion == "-":
            full_version = version
            full_version = full_version.replace("-ce","")
            full_version = full_version.replace("-","")
        else:
            for r in REPLACE_CPE:
                if r in subversion:
                    # if (r + ".") not in subversion:
                        subversion = subversion.replace(r, (r + "."))
            full_version = version + "." + subversion
            full_version = full_version.replace("-ce","")
            full_version = full_version.replace("-","")
    else:
        continue
    for url in GIT_LIST:
        orig_url = url
        url = url.lower()
        try:
            if len(url.split("/")) >= 7:
                url_version = url.split("/")[7]
        except:
            url_version = url.split("/")[-1]
        replaceThis = ["release-","release_", ".release", ".final", "_", "v", ".js", ".zip", ".exe","-ce",".tar.gz", ".dmg", ".bin", ".win-amd64-py2.7",".win32-py2.7", ".win.x64.portable", ".win.x64.", ".tar.bz2", ".win32", ".win64", ".win.x86.portable", ".win.x86", "linux-", "e2fsprogs.", "hdf5.", "windows", "elasticsearch-oss.", "elasticsearch.",  "platform.", ".release", "xwikiplatform", "elefant.", ".stable", "rrrrc.", "betaaaa", "rrc", ".msi", ".tar.xz", ".x64", ".x32", ".tgz",]
        withThis = ["","","","",".","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""]

        try:
            if "github" in url:
                git_component = url.split("/")[4]
                url_version = url_version.replace((git_component+"-"),"")
            elif "wordpress" in url:
                url_version = url_version.replace((url_version.split(".")[0]+"."),"")
        except:
            continue

        for i in range(0,len(replaceThis)):
            url_version = url_version.replace(replaceThis[i],withThis[i])

        replaceThisComponent = [manual_component,"-"]
        withThisComponent = ["","."]
        for i in range(0,len(replaceThisComponent)):
            url_version = url_version.replace(replaceThisComponent[i],withThisComponent[i])
        # print(url_version)    
        for r in REPLACE_GIT:
            if r in url_version:
                if (r + ".") not in url_version:
                    url_version = url_version.replace(r, (r + "."))
                if ("." + r ) not in url_version:
                    url_version = url_version.replace(r, ("." + r ))  
        unusual_test = url_version             
        unusual_test = unusual_test.replace("alpha","")
        unusual_test = unusual_test.replace("beta","")
        # print(unusual_test)
        if "a" in unusual_test or "b" in unusual_test:
            for i in range(0,len(UNUSUAL_GIT)):
                url_version = url_version.replace(UNUSUAL_GIT[i],REPLACE_GIT[i])
                for r in REPLACE_GIT:
                    if r in url_version:
                        if (r + ".") not in url_version:
                            url_version = url_version.replace(r, (r + "."))
                        if ("." + r ) not in url_version:
                            url_version = url_version.replace(r, ("." + r ))  
        if full_version == url_version:
            if orig_url[-1] != "/":
                CPE_URL.append(cpe + orig_url)
                CPE_ARRAY.append(cpe)
        # print(full_version+" | "+url_version)            
    if cpe not in CPE_NOURL and cpe not in CPE_ARRAY:
        CPE_NOURL.append(cpe)
for cpe_url in CPE_URL:
    print(cpe_url + "|" + LICENSE)

for peaseant_cpe in CPE_NOURL:
    print(peaseant_cpe + "NOURL|" + LICENSE)    